api/v1alpha1/authorization_types.go

```diff
@@ -0,0 +1,54 @@
+// Copyright Envoy Gateway Authors
+// SPDX-License-Identifier: Apache-2.0
+// The full text of the Apache license is available in the LICENSE file at
+// the root of the repo.
+
+package v1alpha1
+
+// Authorization defines the authorization configuration.
+type Authorization struct {
+	// Rules contains all the authorization rules.
+	// If rules contains at least one Allow rule and none of them
+	// matches the action for the request is deny.
+	// If rules contains at least one Deny rule and none of them
+	// matches the action for the request is allow.
+	// If rules contains Allow and Deny,
+	// and the request matches both, `Deny` takes precedence.
+	//
+	// +kubebuilder:validation:MinItems=1
+	Rules []Rule `json:"rules,omitempty"`
+}
+
+// Rule defines the single authorization rule.
+type Rule struct {
+	// ClientSelector contains the client selector configuration.
+	// All selectors are AND'd together and only if all selector are valid
+	// the Action is performed.
+	//
+	ClientSelector ClientSelector `json:"clientSelector,omitempty"`
```

Comment:
Should we add another layer here so we can enforce authz on methods?

```
type Rule struct {
	// Action defines the action to be taken if the rule matches.
	Action RuleActionType `json:"action"`

	// Policies define which principals are allowed/denied access to specified permissions based on the action.
	Policies []AuthorizationPolicy `json:"policy"`
}

// AuthorizationPolicy defines the configuration for an authorization policy.
type AuthorizationPolicy struct {
	// Principals define the list of principals that are allowed/denied access based on the action.
	// If no principals are specified, all requests are matched.
	Principles []Principle `json:"principles"`

	// Permissions define the list of permissions that are allowed/denied access based on the action.
	// If no permissions are specified, all requests are matched.
	Permissions []Permission `json:"permissions"`
}
// Permission defines actions that the authorized principals can take.
type Permission struct {
	// Methods define the list of HTTP methods that are allowed/denied access based on the action.
	// If no methods are specified, all methods are matched.
	Method []string `json:"method"`
}
```

Reply:
imo lets create the base structure and add `methods` in a follow up PR, its another attribute like IP
```
rules:
 clientSelector:
   methods:
   - GET
   clientCIDRs:
   - 1.0.0.0/8
 action: "Allow" 
```

Reply:
I still feel like we should use `Principal` here, [it also aligns with Envoy's term](https://www.envoyproxy.io/docs/envoy/latest/api-v3/config/rbac/v3/rbac.proto#envoy-v3-api-msg-config-rbac-v3-principal).

And methods should be in another `permission` field.

```
rules:
   principal:
     clientCIDRs: [1.0.0.0/8]
   permission: [GET]
   action: "Allow"  
```

Reply:
```
rules:
- principal:
     clientCIDRs: [1.0.0.0/8]
     # header: userid // extract user identity from an HTTP header
     # jwtClaim: sub // extract user identity from a jwtClaim
     # other means: client cert, basic auth ...
     # any: true // matches to all principals?
   permission:
     mergemethod: and // supported values and, not, needs better name
     rules:
     - method: GET
       path:
         prefix: /foobar // could support exact, prefix, suffix, contains?
   action: "Allow"
```

so 3 things:
1) permissions should contain way to define path
2) principals may need any: true option which matches all requests? Like if I want to allow all requests to /public path
3) in permissions we may need to have also NOT use case. Like if I want to allow all requests to application but deny /admin path. In that case I would set principals any: true and permissions: mergemethod: not, rules: path: prefix: /admin

@zhaohuabing @arkodg

Reply:
> 1) permissions should contain way to define path

The targeted HTTPRoute would be enough for most cases? I suggest we start with method in the first run.

> 2) principals may need any: true option which matches all requests? Like if I want to allow all requests to /public path

We can make API simpler by allowing `principal` to be optional, which means any.

Reply:
`subject` over `clientSelector` sgtm, can we address these changes as part of a new PR ?

Reply:
`subject` is good. Another new PR is also OK.